File: lua/6-funkcje.lua
#!/bin/lua

function dodaj(a, b)
	return a+b
end

function echo(e)
	print("echo:", e)
	e=e-1
	if e > 0
	then 
		echo(e) 
	end
end

function zwieksz(co, to)
	print("zwieksz: dostałem co=", co, "to=", to)
	co=co+1
	to=to+1
	print("zwieksz: oddaję co=", co, "to=", to)
	return co, to
end

a=2
b=3

print(a, "+", b, "=", dodaj(a, b))

echo(10)

operacja=dodaj;
print("wynik operacji na", a, "i", b, "=", operacja(a,b))

operacja1 = function(a, b)
	return a-b;
end

print("operacja1 odejmij na", a, "i", b,"=", operacja1(a,b))

print("a=", a, "b=", b)
a, b = zwieksz(a, b)
print("a po zwiekszeniu=", a)
print("b po zwiekszeniu=", b)

foo=function(f)
	print("function foo: f=", f)
	if (type(f) == "table") 
	then
		print("f to tablica")
	else
		print("f to ", type(f))
	end
end

foo(123)
foo {456}
foo({789})
print("foo type=", type(foo))

File: perl/6-funkcje.pl
#!/bin/perl

use feature 'state';

sub hello {
	print "hello";
}

sub dodaj
{ # @_ ma argumenty funkcji
	print "dodaj: ilość parametrów=", scalar @_, "->  @_[0] @_[1]\n";
	foreach $el (@_) {
		print "arg -> $el\n";
	}
	return @_[0]+@_[1];
}

sub echo
{
	state $count=0;
	$e=@_[0];
	print "echo e=$e count=$count\n"; 
	$e--;
	$count++;	
	if ($e > 0) {echo($e); }
}

sub pokaz
{
	print "pokaz: private=$private loc=$loc\n";
}

sub zwieksz
{	
	foreach $el (@_) {
		print "zwieksz: arg -> $el\n";
	}
	my $private=@_[0];
	local $loc=@_[0];

	@_[0]++;
	@_[1]++;

	foreach $el (@_) {
		print "zwiekszone: arg -> $el\n";
	}
	print "zwieksz private=$private\n";
	pokaz();
}

hello();

$a=2;
$b=3;

print "$a + $b =", dodaj($a, $b), "\n",;
echo(10);

sub lista {
	my @l=@_; # local variable
	print "lista: parametry: @l\n";
}
lista(10, (1,2,3,4));

sub hash_tab {
	my %h = @_;
	foreach my $key (keys %h) {
		print "key=$key value=$h{$key}\n";
	}
}

%h=('klucz' => 'value', 'k1' => 10);
hash_tab(%h);


$operacja=\&dodaj;
print "wynik operacji na $a i $b=", &$operacja($a,$b), "\n";

print "a = $a b=$b\n";
zwieksz($a, $b);
print "a po zwiekszeniu= $a\n";
print "b po zwiekszeniu= $b\n";
print "global private=$private loc=$loc\n";
